### PR TITLE
Faster and better Sentiment Analysis example. (85% test accuracy in 2 epoches)

### PR DESCRIPTION
**Pull request by farizrahman4u**
Sentiment analysis using recurrent convolutional network. Better and faster results when compared to using either convnet or rnn alone.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
Comparison with other sentiment analysis examples (run on a slow machine so that the time differences are visible):

imdb_lstm.py

Train...
Train on 20000 samples, validate on 5000 samples
Epoch 1/4
20000/20000 [==============================] - 784s - loss: 0.4773 - acc: 0.7769 - val_loss: 0.3613 - val_acc: 0.8396
Epoch 2/4
20000/20000 [==============================] - 788s - loss: 0.2691 - acc: 0.8946 - val_loss: 0.3644 - val_acc: 0.8376
Epoch 3/4
20000/20000 [==============================] - 791s - loss: 0.1770 - acc: 0.9351 - val_loss: 0.3913 - val_acc: 0.8370
Epoch 4/4
20000/20000 [==============================] - 800s - loss: 0.1137 - acc: 0.9612 - val_loss: 0.4621 - val_acc: 0.8308
5000/5000 [==============================] - 48s
Test score: 0.46212145137
Test accuracy: 0.8308

imdb_cnn.py

Train on 20000 samples, validate on 5000 samples
Epoch 1/3
20000/20000 [==============================] - 1414s - loss: 0.6401 - acc: 0.5930 - val_loss: 0.5144 - val_acc: 0.7442
Epoch 2/3
20000/20000 [==============================] - 1411s - loss: 0.3908 - acc: 0.8255 - val_loss: 0.3615 - val_acc: 0.8344
Epoch 3/3
20000/20000 [==============================] - 1416s - loss: 0.3173 - acc: 0.8636 - val_loss: 0.3788 - val_acc: 0.8256



imdb_cnn_lstm.py

Train...
Train on 20000 samples, validate on 5000 samples
Epoch 1/2
20000/20000 [==============================] - 575s - loss: 0.4312 - acc: 0.7900 - val_loss: 0.3457 - val_acc: 0.8456
Epoch 2/2
20000/20000 [==============================] - 580s - loss: 0.2302 - acc: 0.9094 - val_loss: 0.3546 - val_acc: 0.8498
5000/5000 [==============================] - 26s
Test score: 0.354624111649
Test accuracy: 0.8498

File: examples/imdb_cnn_lstm.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+from keras.utils import np_utils
+from keras.models import Sequential
+from keras.layers.core import Dense, Dropout, Activation
+from keras.layers.embeddings import Embedding
+from keras.layers.recurrent import LSTM, GRU, SimpleRNN
+from keras.layers.convolutional import Convolution1D, MaxPooling1D
+from keras.datasets import imdb
+
+'''
+    Train a combination of Convoloution1D and LSTM on the IMDB sentiment classification task.
```
Comment (fchollet) [inline examples/imdb_cnn_lstm.py]: "Convolution". By the way "combination" is somewhat unclear since it first made me think of an ensemble. A "recurrent convolutional network" might be clearer.

File: examples/imdb_cnn_lstm.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    Better results compared to CNN and LSTM examples
```
Comment (fchollet) [inline examples/imdb_cnn_lstm.py]: You can remove this and just specify: 
- accuracy reached, in how many epochs
- time per epoch, on what type of GPU

File: examples/imdb_cnn_lstm.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+maxlen = 100 
+embedding_size=128
+
+#Convolution
+filter_length = 3
+nb_filter = 64
+pool_length=2
+
+#LSTM
+lstm_output_size=70
```
Comment (fchollet) [inline examples/imdb_cnn_lstm.py]: Spaces around `=` throughout the file

**Comment by farizrahman4u** [comment]
@fchollet  Done

**Comment by sergeyf** [comment]
This is pretty awesome, thanks. Is there any way to add masking to the embedding layer to allow variable length sequences?

File: examples/imdb_cnn_lstm.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+'''
+
+
+print("Loading data...")
+(X_train, y_train), (X_test, y_test) = imdb.load_data(nb_words=max_features, test_split=0.2)
+print(len(X_train), 'train sequences')
+print(len(X_test), 'test sequences')
+
+print("Pad sequences (samples x time)")
+X_train = sequence.pad_sequences(X_train, maxlen=maxlen)
```
Comment (farizrahman4u) [inline examples/imdb_cnn_lstm.py]: @sergeyf  X_train and X_test are already padded!

Comment (sergeyf) [inline examples/imdb_cnn_lstm.py]: @farizrahman4u Right, I see that, but if they are padded, doesn't there need to be Masking so that Embedding knows not to pass the zeros on? I tried to add that, but it seems that Convolution1D is not yet compatible with Masking. Do you know what the effect of the padded zeros is on training in this example? 

Thanks.

Comment (farizrahman4u) [inline examples/imdb_cnn_lstm.py]: @sergeyf  As maxlen = 100, which less than the number of words in most of the movie reviews in the dataset, most of X_train and X-test are not padded. So masking will have no effect. In fact, the sentences are actually being trimmed because only 100 words are allowed per sentence. But if you are using a different dataset, such as twitter sentiment analysis datasets, you would have to set maxlen = 140, but most of the tweets would be about less than 14 words long.. and padding would have a significant effect.

Comment (sergeyf) [inline examples/imdb_cnn_lstm.py]: Thank you @farizrahman4u !
